akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2196,6 +2198,97 @@ private[stream] object Collect {
   override def toString: String = "RecoverWith"
 }
 
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class StatefulMap[S, In, Out](create: () => S, f: (S, In) => (S, Out), onComplete: S => Option[Out])
```

Comment:
Wouldn't it be better to add the `onComplete` feature to `statefulMapConcat`. Now those two are somewhat inconsistent. Also, if adding that to `statefulMapConcat` we could implement `statefulMap` with `statefulMapConcat` and wouldn't need another stage implementation.

Reply:
due to the state itself doesn't present in `FlatMapConcat`(it's live in the closure),i don't think it's possible to impl the `onComplete` feature without change current api(since you just can't grab state when using f: () => Out => T)
but i do agree there maybe better ops hierarchy

Reply:
If run on the blocking/io-pool this new operator also makes room for something similar to `unfoldResource` but for mapping with a resource, we discussed adding a factory for that as well, pre-set with a blocking pool, but decided to leave it for another follow up PR.